gpm_core/src/lockfile.rs

```diff
@@ -0,0 +1,92 @@
+use anyhow::Context;
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use std::fs::File;
+use std::io::{Read, Write};
+use std::path::{Path, PathBuf};
+
+/// Allow to look for a specific version of a dependancies
+#[derive(Serialize, Deserialize, Clone, PartialEq, Debug)]
+#[serde(tag = "type")]
+pub enum LockSource {
+    /// Use the mod with the given ID and the matching version.
+    ///
+    /// Even if the parent contain the id, we still store it there, as it may be usefull that id
+    /// differ from it's parent.
+    IdVersion { identifier: String, version: String },
+    /// Use a specific path on the local filesystem. If relative, it'll be based around the
+    /// profile folder.
+    Path { path: PathBuf },
+}
+
+/// contain a fixed set of mod depency, with each depency having a specific version. Mod are
+/// identified by their id. They are unique.
+#[derive(Serialize, Deserialize, Default, Debug)]
+pub struct LockFile {
+    pub dependancies: HashMap<String, LockSource>,
+}
+
+impl LockFile {
+    /// create a new [`LockFile`] with no dependancies.
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    /// return the [`LockSource`] corresponding to a given package identifier if it is present in
+    /// this [`LockFile`], None otherwise.
+    pub fn depency_source(&self, identifier: &str) -> Option<LockSource> {
+        self.dependancies.get(identifier).cloned()
+    }
+
+    /// define the [`LockSource`] for a given package identifier, overwriting the current one.
+    ///
+    /// return the previous [`LockSource`] if overwriting it
+    pub fn set_depency_source(
+        &mut self,
+        identifier: String,
+        source: LockSource,
+    ) -> Option<LockSource> {
+        self.dependancies.insert(identifier, source)
+    }
+
+    //TODO: do we want to store them as JSON or as TOML ?
```

Comment:
It should probably use the same language we use for the package or project to be able to use the same parsing library

Reply:
If we follow the specification, we will in anyway need both json and toml. I personally prefer toml, but the current state of discussion is that we use toml for human edited content and json for machine generated content.

Reply:
We can consider our gpm CLI tool to only use TOML for everything. The format it builds packages into however is not the responsibility of gpm to specify, it's merely the target it follows.

Reply:
I see, you plan to use json only for the standard for easy interability with other tool. It's clear now.